shadowsocks/dialer_test.go

```diff
@@ -0,0 +1,212 @@
+package shadowsocks
+
+import (
+	"io"
+	"net"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-ss-server/metrics"
+	onet "github.com/Jigsaw-Code/outline-ss-server/net"
+	"github.com/shadowsocks/go-shadowsocks2/shadowaead"
+	"github.com/shadowsocks/go-shadowsocks2/socks"
+)
+
+const (
+	testCipher   = "chacha20-ietf-poly1305"
+	testPassword = "testPassword"
+	testPayload  = "!!!test~payload!!!"
+)
+
+func TestShadowsocksDialer_DialTCP(t *testing.T) {
+	// Let the OS pick an available port
+	listenTCPAddr := &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0}
+	proxyListener, err := net.ListenTCP("tcp", listenTCPAddr)
+	if err != nil {
+		t.Fatalf("ListenTCP failed: %v", err)
+	}
+	go startShadowsocksTCPProxy(proxyListener, t)
+
+	targetListener, err := net.ListenTCP("tcp", listenTCPAddr)
+	go startTCPEchoServer(targetListener, t)
+
+	proxyHost, proxyPortStr, err := net.SplitHostPort(proxyListener.Addr().String())
+	if err != nil {
+		t.Fatalf("Failed to parse proxy address: %v", err)
+	}
+	proxyPort, _ := strconv.Atoi(proxyPortStr)
+	d, err := NewDialer(proxyHost, testPassword, testCipher, proxyPort)
+	if err != nil {
+		t.Fatalf("Failed to create ShadowsocksDialer: %v", err)
+	}
+	conn, err := d.DialTCP(targetListener.Addr().String())
+	if err != nil {
+		t.Fatalf("ShadowsocksDialer.DialTCP failed: %v", err)
+	}
+	expectEchoPayload(conn, len(testPayload), t)
+}
+
+func TestShadowsocksDialer_DialUDP(t *testing.T) {
+	// Let the OS pick an available port
+	listenUDPAddr := &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0}
+	proxyConn, err := net.ListenUDP("udp", listenUDPAddr)
+	if err != nil {
+		t.Fatalf("Proxy ListenUDP failed: %v", err)
+	}
+	go startShadowsocksUDPProxy(proxyConn, t)
+
+	targetConn, err := net.ListenUDP("udp", listenUDPAddr)
+	if err != nil {
+		t.Fatalf("Target ListenUDP failed: %v", err)
+	}
+	go startUDPEchoServer(targetConn, t)
+
+	proxyHost, proxyPortStr, err := net.SplitHostPort(proxyConn.LocalAddr().String())
+	if err != nil {
+		t.Fatalf("Failed to parse proxy address: %v", err)
+	}
+	proxyPort, _ := strconv.Atoi(proxyPortStr)
+	d, err := NewDialer(proxyHost, testPassword, testCipher, proxyPort)
+	if err != nil {
+		t.Fatalf("Failed to create ShadowsocksDialer: %v", err)
+	}
+	conn, err := d.DialUDP(targetConn.LocalAddr().String())
+	if err != nil {
+		t.Fatalf("ShadowsocksDialer.DialUDP failed: %v", err)
+	}
+	conn.SetReadDeadline(time.Now().Add(time.Second * 5))
+	expectEchoPayload(conn, udpBufSize, t)
+}
+
+func startShadowsocksTCPProxy(listener *net.TCPListener, t *testing.T) {
```

Comment:
You effectively re-implementing the server functionality here. Same with UDP.
I wonder if we want to make it more of a unit test. There's no need to do a Dial to the target in order to test the client. All you need is to specify the response and save the target address and the connection payload, then check if the client received the response correctly and that it passed the address and payload correctly.

Reply:
I have implemented Shadowsocks TCP and UDP echo proxies that validate a target address.